File: openfilz-sdk/openfilz-sdk-python/samples/quickstart.py
"""
OpenFilz Python SDK Quick Start

Demonstrates the core document management operations:
- Folder creation, listing
- File upload, download, rename, move, copy, delete
- Favorites, metadata, dashboard, audit trail

Prerequisites:
    - A running OpenFilz API instance (default: http://localhost:8081)
    - pip install openfilz-sdk-python
"""

import os
import sys
import tempfile

import openfilz_sdk
from openfilz_sdk.api import (
    document_controller_api,
    folder_controller_api,
    file_controller_api,
    favorites_api,
    dashboard_api,
    audit_controller_api,
)
from openfilz_sdk.models.create_folder_request import CreateFolderRequest
from openfilz_sdk.models.rename_request import RenameRequest
from openfilz_sdk.models.copy_request import CopyRequest
from openfilz_sdk.models.move_request import MoveRequest
from openfilz_sdk.models.delete_request import DeleteRequest
from openfilz_sdk.models.update_metadata_request import UpdateMetadataRequest


def run_quickstart():
    api_url = os.environ.get("OPENFILZ_API_URL", "http://localhost:8081")

    # ──── 1. Configure the SDK client ────────────────────────────────
    configuration = openfilz_sdk.Configuration(host=api_url)
    # For authenticated instances:
    # configuration.access_token = "your-oauth2-access-token"

    with openfilz_sdk.ApiClient(configuration) as api_client:
        folder_api = folder_controller_api.FolderControllerApi(api_client)
        document_api = document_controller_api.DocumentControllerApi(api_client)
        file_api = file_controller_api.FileControllerApi(api_client)
        fav_api = favorites_api.FavoritesApi(api_client)
        dashboard = dashboard_api.DashboardApi(api_client)
        audit_api = audit_controller_api.AuditControllerApi(api_client)

        # ──── 2. Create a folder ─────────────────────────────────────────
        folder = folder_api.create_folder(
            CreateFolderRequest(name="Python Quick Start")
        )
        folder_id = folder.id
        print(f"Created folder: {folder.name} (id={folder_id})")

        # ──── 3. Upload a file ───────────────────────────────────────────
        with tempfile.NamedTemporaryFile(
            mode="w", suffix=".txt", delete=False
        ) as tmp:
            tmp.write("Hello from OpenFilz Python SDK!")
            tmp_path = tmp.name

        try:
            uploaded = document_api.upload_document1(
                file=tmp_path,
                allow_duplicate_file_names=True,
                parent_folder_id=str(folder_id),
            )
            file_id = uploaded.id
            print(f"Uploaded: {uploaded.name} (id={file_id})")

            # ──── 4. List folder contents ────────────────────────────────────
            contents = folder_api.list_folder(
                folder_id=folder_id, only_files=False, only_folders=False
            )
            print(f"Folder contains {len(contents)} item(s):")
            for item in contents:
                print(f"  {item.type}: {item.name}")

            # ──── 5. Get document info ───────────────────────────────────────
            info = document_api.get_document_info(file_id, with_metadata=True)
            print(f"Document: name={info.name}, type={info.content_type}")

            # ──── 6. Update metadata ─────────────────────────────────────────
            document_api.update_document_metadata(
                file_id,
                UpdateMetadataRequest(
                    metadata_to_update={"project": "Python Demo", "version": 1}
                ),
            )
            print("Metadata updated")

            # ──── 7. Download the file ───────────────────────────────────────
            downloaded = document_api.download_document(file_id)
            print(f"Downloaded file: {downloaded}")

            # ──── 8. Rename the file ─────────────────────────────────────────
            renamed = file_api.rename_file(
                file_id, RenameRequest(new_name="py-renamed.txt")
            )
            print(f"Renamed to: {renamed.name}")

            # ──── 9. Create target folder and copy ───────────────────────────
            target_folder = folder_api.create_folder(
                CreateFolderRequest(name="Python Quick Start Target")
            )
            target_folder_id = target_folder.id

            copies = file_api.copy_files(
                CopyRequest(
                    document_ids=[file_id], target_folder_id=target_folder_id
                )
            )
            copy_id = copies[0].copy_id
            print(f"Copied file, new id={copy_id}")

            # ──── 10. Move the original file ─────────────────────────────────
            file_api.move_files(
                MoveRequest(
                    document_ids=[file_id],
                    target_folder_id=target_folder_id,
                    allow_duplicate_file_names=True,
                )
            )
            print("Moved file to target folder")

            # ──── 11. Toggle favorite ────────────────────────────────────────
            is_favorite = fav_api.toggle_favorite(file_id)
            print(f"Favorite toggled: {is_favorite}")

            # ──── 12. Dashboard statistics ───────────────────────────────────
            stats = dashboard.get_dashboard_statistics()
            print(
                f"Dashboard: files={stats.total_files}, folders={stats.total_folders}"
            )

            # ──── 13. Audit trail ────────────────────────────────────────────
            trail = audit_api.get_audit_trail(file_id, sort="DESC")
            print(f"Audit trail: {len(trail)} entries")

            # ──── 14. Cleanup ────────────────────────────────────────────────
            fav_api.remove_favorite(file_id)
            file_api.delete_files(DeleteRequest(document_ids=[file_id, copy_id]))
            folder_api.delete_folders(
                DeleteRequest(document_ids=[folder_id, target_folder_id])
            )
            print("Cleanup complete")

        finally:
            os.unlink(tmp_path)


if __name__ == "__main__":
    run_quickstart()

File: openfilz-sdk/openfilz-sdk-typescript/samples/quickstart.ts
/**
 * OpenFilz TypeScript SDK Quick Start
 *
 * Demonstrates the core document management operations:
 * - Folder creation, listing
 * - File upload, download, rename, move, copy, delete
 * - Favorites, metadata, dashboard, audit trail
 *
 * Prerequisites:
 *   - A running OpenFilz API instance (default: http://localhost:8081)
 *   - npm install @openfilz-sdk/typescript
 */

import {
  Configuration,
  DocumentControllerApi,
  FolderControllerApi,
  FileControllerApi,
  FavoritesApi,
  DashboardApi,
  AuditControllerApi,
} from '@openfilz-sdk/typescript';
import * as fs from 'fs';
import * as fsp from 'fs/promises';
import * as path from 'path';
import { pipeline } from 'stream/promises';

const API_URL = process.env.OPENFILZ_API_URL || 'http://localhost:8081';

export async function runQuickStart(): Promise<void> {
  // ──── 1. Configure the SDK client ────────────────────────────────
  const config = new Configuration({
    basePath: API_URL,
    // For authenticated instances:
    // accessToken: 'your-oauth2-access-token',
  });

  const folderApi = new FolderControllerApi(config);
  const documentApi = new DocumentControllerApi(config);
  const fileApi = new FileControllerApi(config);
  const favApi = new FavoritesApi(config);
  const dashboardApi = new DashboardApi(config);
  const auditApi = new AuditControllerApi(config);

  // ──── 2. Create a folder ─────────────────────────────────────────
  const folder = await folderApi.createFolder({
    name: 'TypeScript Quick Start',
    parentId: undefined,
  });
  const folderId = folder.data.id;
  console.log(`Created folder: ${folder.data.name} (id=${folderId})`);

  // ──── 3. Upload a file ───────────────────────────────────────────
  const tempFilePath = path.join(__dirname, 'temp-quickstart.txt');
  await fsp.writeFile(tempFilePath, 'Hello from OpenFilz TypeScript SDK!');

  // Async file read — avoids blocking the event loop
  const fileBuffer = await fsp.readFile(tempFilePath);
  const uploaded = await documentApi.uploadDocument1(
    new File([fileBuffer], 'quickstart-demo.txt', { type: 'text/plain' }),
    true,       // allowDuplicateFileNames
    folderId,   // parentFolderId
    undefined,  // metadata
  );
  const fileId = uploaded.data.id!;
  console.log(`Uploaded: ${uploaded.data.name} (id=${fileId})`);

  // ──── 4. List folder contents ────────────────────────────────────
  const contents = await folderApi.listFolder(folderId, false, false);
  console.log(`Folder contains ${contents.data.length} item(s):`);
  contents.data.forEach((item) => {
    console.log(`  ${item.type}: ${item.name}`);
  });

  // ──── 5. Get document info ───────────────────────────────────────
  const info = await documentApi.getDocumentInfo(fileId, true);
  console.log(`Document: name=${info.data.name}, type=${info.data.contentType}`);

  // ──── 6. Update metadata ─────────────────────────────────────────
  await documentApi.updateDocumentMetadata(fileId, {
    metadataToUpdate: { project: 'TypeScript Demo', version: 1 },
  });
  console.log('Metadata updated');

  // ──── 7. Download the file ───────────────────────────────────────
  // Stream to a temp file instead of buffering entire response in memory
  const downloadPath = path.join(__dirname, 'temp-download.txt');
  const downloadResp = await documentApi.downloadDocument(fileId, {
    responseType: 'stream',
  });
  await pipeline(downloadResp.data as any, fs.createWriteStream(downloadPath));
  const downloadedSize = (await fsp.stat(downloadPath)).size;
  console.log(`Downloaded file to disk (${downloadedSize} bytes)`);
  await fsp.unlink(downloadPath);

  // ──── 8. Rename the file ─────────────────────────────────────────
  const renamed = await fileApi.renameFile(fileId, { newName: 'ts-renamed.txt' });
  console.log(`Renamed to: ${renamed.data.name}`);

  // ──── 9. Create target folder and copy the file ──────────────────
  const targetFolder = await folderApi.createFolder({
    name: 'TypeScript Quick Start Target',
  });
  const targetFolderId = targetFolder.data.id;

  const copies = await fileApi.copyFiles({
    documentIds: [fileId],
    targetFolderId: targetFolderId,
  });
  const copyId = copies.data[0].copyId!;
  console.log(`Copied file, new id=${copyId}`);

  // ──── 10. Move the original file ─────────────────────────────────
  await fileApi.moveFiles({
    documentIds: [fileId],
    targetFolderId: targetFolderId,
    allowDuplicateFileNames: true,
  });
  console.log('Moved file to target folder');

  // ──── 11. Toggle favorite ────────────────────────────────────────
  const isFavorite = await favApi.toggleFavorite(fileId);
  console.log(`Favorite toggled: ${isFavorite.data}`);

  // ──── 12. Dashboard statistics ───────────────────────────────────
  const stats = await dashboardApi.getDashboardStatistics();
  console.log(`Dashboard: files=${stats.data.totalFiles}, folders=${stats.data.totalFolders}`);

  // ──── 13. Audit trail ────────────────────────────────────────────
  const trail = await auditApi.getAuditTrail(fileId, 'DESC');
  console.log(`Audit trail: ${trail.data.length} entries`);

  // ──── 14. Cleanup ────────────────────────────────────────────────
  await favApi.removeFavorite(fileId);
  await fileApi.deleteFiles({ documentIds: [fileId, copyId] });
  await folderApi.deleteFolders({ documentIds: [folderId, targetFolderId] });
  console.log('Cleanup complete');

  // Remove temp file
  await fsp.unlink(tempFilePath).catch(() => {});
}

// Run if executed directly
if (require.main === module) {
  runQuickStart().catch(console.error);
}
